### PR TITLE
Add ByteArrayToImageSourceConverter Sample

### LINKED ISSUE #51: [Proposal] `byte[]`-to-`ImageSource` Converter
# ByteArrayToImageSourceConverter

* [x] Proposed
* [x] Prototype
* [x] Implementation
  * [x] iOS Support
  * [x] Android Support
  * [x] macOS Support
  * [x] Windows Support
* [x] Unit Tests
* [x] Sample
* [x] Documentation

## Summary
[summary]: #summary

The `ByteArrayToImageSourceConverter` is a `Converter` that allows the user to convert an incoming value from `byte[]` and returns an `ImageSource`. This object can then be used as the Source of an Image control.

## Detailed Design
[design]: #detailed-design

## Usage Syntax
[usage]: #usage-syntax

### XAML Usage

```xml
<ContentPage xmlns="http://xamarin.com/schemas/2014/forms"
             xmlns:x="http://schemas.microsoft.com/winfx/2009/xaml"
             xmlns:xct="http://xamarin.com/schemas/2020/toolkit"
             x:Class="MyLittleApp.MainPage">
    <ContentPage.Resources>
         <ResourceDictionary>
             <xct:ByteArrayToImageSourceConverter x:Key="ByteArrayToImageSourceConverter" />
         </ResourceDictionary>
    </ContentPage.Resources>

    <StackLayout>

        <Image Source="{Binding MyByteArray, Converter={StaticResource ByteArrayToImageSourceConverter}}" />

    </StackLayout>
</ContentPage>
```

### C# Usage

```cs
class MyPage : ContentPage
{
  public MyPage()
  {
    Content = new Image().Bind(Image.SourceProperty, named(ViewModel.MyByteArray), converter: new ByteArrayToImageSourceConverter());
  }
}
```

### PR DESCRIPTION
**Pull request by brminnick**
### Description of Change ###

This PR adds `ByteArrayToImageSourceConverterPage` to `CommunityToolkit.Maui.Sample`.

While implementing `ByteArrayToImageSourceConverterPage`, I discovered that the following pages were incorrectly not using `BasePage<TViewModel>` and not using Dependency Injection, and I've updated them accordingly:
- `SnackbarPage`
- `ToastPage`
- `CharactersValidationBehaviorPage`
- `EmailValidationBehaviorPage`
- `InvertedBoolConverterPage`
- `MaskedBehaviorPage`
- `MultiValidationBehaviorPage`
- `NumericValidationBehaviorPage`
- `RequiredStringValidationBehaviorPage`
- `SetFocusOnEntryCompletedBehaviorPage`
- `TextValidationBehaviorPage`
- `UriValidationBehaviorPage`
- `BoolToObjectConverterPage`
- `ColorsConverterPage`
- `EnumToBoolConverterPage`
- `IsListNotNullOrEmptyConverterPage`
- `IsListNullOrEmptyConverterPage`
- `MathExpressionConverterPage`
- `MultiConverterPage`
- `NotEqualConverterPage`
- `StringToListConverterPage`
- `TextCaseConverterPage`

To ensure future samples added to `CommunityToolkit.Maui.Sample` implement `BasePage<TViewModel>` properly, I created the following method to be used when initializing `IReadOnlyDictionary AppShell.viewModelMaddings`:

```cs
static KeyValuePair<Type, (Type GalleryPageType, Type ContentPageType)> CreateViewModelMapping<TPage, TViewModel, TGalleryPage, TGalleryViewModel>() 
  where TPage : BasePage<TViewModel>
  where TViewModel : BaseViewModel
  where TGalleryPage : BaseGalleryPage<TGalleryViewModel>
  where TGalleryViewModel : BaseGalleryViewModel
{
	return new KeyValuePair<Type, (Type GalleryPageType, Type ContentPageType)>(typeof(TViewModel), (typeof(TGalleryPage), typeof(TPage)));
}
```

 ### Linked Issues ###
 

 - Fixes #51 

 ### PR Checklist ###
 
 - [x] Has a linked Issue, and the Issue has been `approved`(bug) or `Championed` (feature/proposal)
 - [x] Has tests (if omitted, state reason in description)
 - [x] Has samples (if omitted, state reason in description)
 - [x] Rebased on top of `main` at time of PR
 - [x] Changes adhere to [coding standard](https://github.com/CommunityToolkit/Maui/blob/main/CONTRIBUTING.md#contributing-code---best-practices)

 ### Additional information ###

### CODE REVIEW

File: samples/CommunityToolkit.Maui.Sample/AppShell.xaml.cs [hunk 1]
```diff
@@ -12,62 +14,63 @@ namespace CommunityToolkit.Maui.Sample;
 
 public partial class AppShell : Shell
 {
-	static readonly IReadOnlyDictionary<Type, (Type GalleryPageType, Type ContentPageType)> viewModelMappings = new Dictionary<Type, (Type, Type)>()
+	static readonly IReadOnlyDictionary<Type, (Type GalleryPageType, Type ContentPageType)> viewModelMappings = new Dictionary<Type, (Type, Type)>(new[]
 	{
 		// Add Alerts View Models
-		{ typeof(SnackbarViewModel), (typeof(AlertsGalleryPage), typeof(SnackbarPage)) },
-		{ typeof(ToastViewModel), (typeof(AlertsGalleryPage), typeof(ToastPage)) },
+		CreateViewModelMapping<SnackbarPage, SnackbarViewModel, AlertsGalleryPage, AlertsGalleryViewModel>(),
```
Comment (VladislavAntonyuk) [inline samples/CommunityToolkit.Maui.Sample/AppShell.xaml.cs:20]: AlertGalleryViewModel is not necessary here. This mapping is used to build the path for Shell

File: samples/CommunityToolkit.Maui.Sample/AppShell.xaml.cs [hunk 2]
```diff
@@ -93,9 +96,17 @@ public static string GetPageRoute(Type viewModelType)
 
 	static void RegisterRouting()
 	{
-		foreach(var viewModelKeyValuePair in viewModelMappings)
+		foreach (var viewModelKeyValuePair in viewModelMappings)
 		{
 			Routing.RegisterRoute(GetPageRoute(viewModelKeyValuePair.Value.GalleryPageType, viewModelKeyValuePair.Value.ContentPageType), viewModelKeyValuePair.Value.ContentPageType);
```
Comment (VladislavAntonyuk) [inline samples/CommunityToolkit.Maui.Sample/AppShell.xaml.cs:98]: You can see TGalleryViewModel is not used here.

File: samples/CommunityToolkit.Maui.Sample/Pages/Converters/NotEqualConverterPage.xaml.cs [hunk 3]
```diff
@@ -1,11 +1,12 @@
...
+
+public partial class NotEqualConverterPage : BasePage<NotEqualConverterViewModel>
 {
-	public NotEqualConverterPage()
+	public NotEqualConverterPage(NotEqualConverterViewModel notEqualConverterViewModel)
+		: base(notEqualConverterViewModel)
 	{
 		InitializeComponent();
-
-		ExampleText ??= new();
```
Comment (VladislavAntonyuk) [inline samples/CommunityToolkit.Maui.Sample/Pages/Converters/NotEqualConverterPage.xaml.cs:9]: Please check if it works on Windows

File: samples/CommunityToolkit.Maui.Sample/Pages/Converters/StringToListConverterPage.xaml [hunk 4]
```diff
@@ -17,29 +21,34 @@
         </ResourceDictionary>
     </pages:BasePage.Resources>
     <pages:BasePage.Content>
-        <Grid Margin="20,0" RowDefinitions="Auto,Auto,*">
+        <Grid Margin="20,0"
```
Comment (VladislavAntonyuk) [inline samples/CommunityToolkit.Maui.Sample/Pages/Converters/StringToListConverterPage.xaml:24]: Can we use VerticalStackLayout?
What has better performance: Grid or VerticalStackLayout?

File: samples/CommunityToolkit.Maui.Sample/ViewModels/Converters/ByteArrayToImageSourceConverterViewModel.cs [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+	async Task ExecuteDownloadDotNetBotImageCommand()
+	{
+		IsDownloadingImage = true;
+
+		// Ensure Activity Indicator appears on screen for a minumum of 1.5 seconds when the uses taps the Download Button
+		var minimumDownloadTimeTask = Task.Delay(TimeSpan.FromSeconds(1.5));
+
+		try
+		{
+			DotNetBotImageByteArray = await client.GetByteArrayAsync("https://user-images.githubusercontent.com/13558917/137551073-ac8958bf-83e3-4ae3-8623-4db6dce49d02.png").ConfigureAwait(false);
```
Comment (VladislavAntonyuk) [inline samples/CommunityToolkit.Maui.Sample/ViewModels/Converters/ByteArrayToImageSourceConverterViewModel.cs:57]: Can we use image from Embedded Resources? To avoid using HttpClient and extra dependency and potential issue if there is no the Internet connection

File: samples/CommunityToolkit.Maui.Sample/Pages/Converters/StringToListConverterPage.xaml (hunk 4 above)
Comment (brminnick) [reply samples/CommunityToolkit.Maui.Sample/Pages/Converters/StringToListConverterPage.xaml:24]: > Can we use VerticalStackLayout?

Totally! I'll update it 👍

File: samples/CommunityToolkit.Maui.Sample/ViewModels/Converters/ByteArrayToImageSourceConverterViewModel.cs (hunk 5 above)
Comment (brminnick) [reply samples/CommunityToolkit.Maui.Sample/ViewModels/Converters/ByteArrayToImageSourceConverterViewModel.cs:57]: Ah, I like injecting `HttpClient! 

I think it's a good example that we can share with devs on how to add `HttpClient` to `IServiceCollection` via `builder.Services.AddHttpClient()`, then how to inject it into a ViewModel.

And the `Microsoft.Extensions.Http` dependency is only added to the sample app, not to the `CommunityToolkit.Maui` or `CommunityToolkit.Maui.Core` libraries.

Good call on the lack of internet connectivity, though! I'll add a `catch` block to ensure the app doesn't crash when there's no internet connectivity 👍

File: samples/CommunityToolkit.Maui.Sample/Pages/Converters/NotEqualConverterPage.xaml.cs (hunk 3 above)
Comment (brminnick) [reply samples/CommunityToolkit.Maui.Sample/Pages/Converters/NotEqualConverterPage.xaml.cs:9]: In this case, I removed `x:Name:ExampleText` from the XAML, and the field `ExampleText` no longer exists. 

Last I checked, the `??= new()` workaround is still required on Windows fro Preview 13. Hopefully that'll be fixed in the RC! 🤞

File: samples/CommunityToolkit.Maui.Sample/AppShell.xaml.cs (hunk 1 above)
Comment (brminnick) [reply samples/CommunityToolkit.Maui.Sample/AppShell.xaml.cs:20]: You're totally right! 

But I couldn't figure out a way to constrain `TGalleryPage` for `CreateViewModelMapping` without also requiring + constraining `TGalleryViewModel `.

The reason I created the `CreateViewModelMapping` method is to ensure that every new PR implements the following correctly:
- [x] Sample Page implements `BasePage<T>`
- [x] Sample View Models implement `BaseViewModel` 
- [x] Sample Gallery Pages implement `BaseGalleryPage<T>` 

To ensure this, I implemented the following constraints on `CreateViewModelMapping`:
- `where TPage : BasePage<TViewModel>`
- `where TViewModel : BaseViewModel`
- `where TGalleryPage : BaseGalleryPage<TGalleryViewModel>`

But in order to constrain `TGalleryPage` to `BaseGalleryPage<T>`, the constraints also needs to know that `T` is constrained to `BaseGalleryViewModel`. 

And when I try to use `where TGalleryPage : BaseGalleryPage<BaseGalleryViewModel>`, I get this error:

| `CreateViewModelMapping` without `TGalleryViewModel` | Error | 
| ---------------------------------------------------------- | ----- |
| Screen Shot 2022-03-05 at 12 08 35 PM | image |


So, long story short, I couldn't figure out a way to constrain `TGalleryPage` without also including `TGalleryViewModel`.

I'm totally down to remove `TGalleryViewModel` if you can figure out how to implement `CreateViewModelMapping` without it, though!

File: samples/CommunityToolkit.Maui.Sample/AppShell.xaml.cs (hunk 2 above)
Comment (brminnick) [reply samples/CommunityToolkit.Maui.Sample/AppShell.xaml.cs:98]: Right! Here's more background + detail on my rationale behind the implementation: https://github.com/CommunityToolkit/Maui/pull/317#discussion_r820143745

File: samples/CommunityToolkit.Maui.Sample/Pages/Converters/StringToListConverterPage.xaml (hunk 4 above)
Comment (brminnick) [reply samples/CommunityToolkit.Maui.Sample/Pages/Converters/StringToListConverterPage.xaml]: Done ✅ 

> What has better performance: Grid or VerticalStackLayout

If the `Grid` uses `GridLength.Auto`, its performance drops significantly to a similar level of `VerticalStackLayout` because it has to resize its row/column height/width based on the size of its `Grid.Children`. AND it has to resize each time a child updates.

If the `Grid` uses only `GridLength.Star` or `GridLength.Absolute`, its performance improves significantly over `VerticalStackLayout`.

**Review by VladislavAntonyuk — APPROVED** [review]

**Comment by brminnick** [comment]
Thanks Vlad!!
